sweep/aggregator.go

```diff
@@ -492,28 +491,51 @@ func NewBudgetAggregator(estimator chainfee.Estimator,
 }
 
 // clusterGroup defines an alias for a set of inputs that are to be grouped.
-type clusterGroup map[fn.Option[int32]][]pendingInput
+type clusterGroup map[int32][]SweeperInput
 
 // ClusterInputs creates a list of input sets from pending inputs.
 // 1. filter out inputs whose budget cannot cover min relay fee.
-// 2. group the inputs into clusters based on their deadline height.
-// 3. sort the inputs in each cluster by their budget.
-// 4. optionally split a cluster if it exceeds the max input limit.
-// 5. create input sets from each of the clusters.
-func (b *BudgetAggregator) ClusterInputs(inputs pendingInputs) []InputSet {
+// 2. filter a list of exclusive inputs.
+// 3. group the inputs into clusters based on their deadline height.
+// 4. sort the inputs in each cluster by their budget.
+// 5. optionally split a cluster if it exceeds the max input limit.
+// 6. create input sets from each of the clusters.
+// 7. create input sets for each of the exclusive inputs.
+func (b *BudgetAggregator) ClusterInputs(inputs InputsMap,
+	defaultDeadline int32) []InputSet {
+
 	// Filter out inputs that have a budget below min relay fee.
 	filteredInputs := b.filterInputs(inputs)
 
 	// Create clusters to group inputs based on their deadline height.
 	clusters := make(clusterGroup, len(filteredInputs))
 
+	// exclusiveInputs is a set of inputs that are not to be included in
+	// any cluster. These inputs can only be swept independently as there's
+	// no guarantee which input will be confirmed first, which means
+	// grouping exclusive inputs may jeopardize non-exclusive inputs.
+	exclusiveInputs := make(map[wire.OutPoint]clusterGroup)
+
 	// Iterate all the inputs and group them based on their specified
 	// deadline heights.
 	for _, input := range filteredInputs {
-		height := input.params.DeadlineHeight
+		// Get deadline height, and use the specified default deadline
+		// height if it's not set.
+		height := input.params.DeadlineHeight.UnwrapOr(defaultDeadline)
+
+		// Put exclusive inputs in their own set.
+		if input.params.ExclusiveGroup != nil {
+			log.Tracef("Input %v is exclusive", input.OutPoint())
+			exclusiveInputs[input.OutPoint()] = clusterGroup{
+				height: []SweeperInput{*input},
+			}
+
+			continue
+		}
```

Comment:
Nice -- in `BudgetAggregator` exclusive groups are now truly "exclusive" in that there's no way for them to be batched with other inputs.

But with all the refactoring, exclusive groups were broken for `SimpleAggregator`.  The exponential backoff was removed, which means invalid anchor spends can be repeatedly clustered with valid spends, preventing the sweep tx from ever being successfully broadcast.

We should either fix that or remove `SimpleAggregator` altogether.

Reply:
If no one has an argument for keeping it, then if it doesn't work we should just 🔪 it.

Reply:
I think it was left in place to make the initial set of commits easier to follow. However, it's still used as the default aggregator here: https://github.com/lightningnetwork/lnd/blob/b9485d3cad0dbe6dd328b45ead42fffd680226ed/server.go#L1066-L1069

I assume this'll be fixed in the next series in this PR.

Reply:
Yep we can remove it in the final PR.